torch_geometric/contrib/nn/models/graph_lottery_ticket.py

```diff
@@ -0,0 +1,515 @@
+    save_all_masks: bool = False
+    seed: int = field(default_factory=lambda: randint(1, 9999))
+    verbose: bool = False
+    ignore_keys: set = field(default_factory=lambda: set())
+
+    def __post_init__(self):
+        """fixes internal hyperparams and generates GLTMask from
+        :obj:`GLTMask` from model """
+        torch.manual_seed(self.seed)
+
+        if not self.lr_mask_graph:
+            self.lr_mask_graph = self.lr
+
+        if not self.lr_mask_model:
+            self.lr_mask_model = self.lr
+
+        self.optim_args = {"lr": self.lr, **self.optim_args}
+
+        self.mask = GLTMask(
+            self.module, self.graph,
+            torch.device("cuda" if torch.cuda.is_available() else "cpu"))
+
+    def prune(self) -> Tuple[Dict[str, Any], Dict[str, Any], Dict[str, float]]:
+        """UGS algorithm. Train model with UGS to train masks and params.
+        Disretize masks by pruning rates. Retrain without UGS for final
+        performance. """
+        initial_params = {
+            "module." + k + GLTModel.ORIG if k.rpartition(".")[-1]
+            not in self.ignore_keys else "module." + k: v.detach().clone()
+            for k, v in self.module.state_dict().items()
+        }
+
+        ticket = GLTModel(self.module, self.graph,
+                          ignore_keys=self.ignore_keys)
+        ticket.apply_mask(self.mask.to_dict())
+
+        unrewound_test_score, masks = self.train(ticket, True)
+        if self.verbose:
+            print("[UNREWOUND] Final test performance:", unrewound_test_score)
+        self.mask.load_and_binarize(masks, self.prune_rate_model,
+                                    self.prune_rate_graph)
+
+        mask_dict = self.mask.to_dict(weight_prefix=True)
+        ticket.rewind({**mask_dict, **initial_params})
+        fixed_test_score, masks = self.train(ticket, False)
+
+        if self.verbose:
+            print("[FIXED MASK] Final test performance:", fixed_test_score)
+        current_sparsity = self.mask.sparsity()
+        if self.verbose:
+            print(
+                "Graph sparsity:",
+                round(current_sparsity[0], 4),
+                "Model sparsity:",
+                round(current_sparsity[1], 4),
+            )
+        results_dict = {
+            "unrewound_test": unrewound_test_score,
+            "fixed_test": fixed_test_score,
+            "graph_sparsity": current_sparsity[0],
+            "model_sparsity": current_sparsity[1]
+        }
+        return initial_params, self.mask.to_dict(), results_dict
+
+    def train(self, ticket: GLTModel,
+              ugs: bool) -> Tuple[float, Dict[str, Parameter]]:
+        """train loop. If ugs flag, use mask regularization."""
+        best_val_score = 0.0
+        final_test_score = 0.0
+        best_masks = {}
+        optimizer = self.optimizer(ticket.parameters(), **self.optim_args)
+
+        with tqdm.trange(self.max_train_epochs, disable=not self.verbose) as t:
```

Comment:
It looks like the parameters of the model might also get updated in this training loop, I guess that's why you have `rewind`. Instead wouldn't calling `self.module.eval()` before running this simply prevent model parameters from getting updated?

Reply:
The LTH relies on *jointly* updating model and mask parameters from their initialization to learn the best subnetwork, so the model should remain in train mode as far as I know.